back/spec/models/user_spec.rb

```diff
@@ -1257,6 +1257,83 @@
     end
   end
 
+  describe '#compacted_roles' do
+    let_it_be(:user) { create(:user) }
+
+    let_it_be(:projects_in_folder) { create_list(:project, 2) }
+    let_it_be(:folder) { create(:project_folder, projects: projects_in_folder) }
+
+    # Top-level project (not in a folder)
+    let_it_be(:another_project) { create(:project) }
+
+    let_it_be(:project_in_another_folder) { create(:project) }
+    let_it_be(:another_folder) { create(:project_folder, projects: [project_in_another_folder]) }
+
+    def create_roles(projects, folders, admin: false)
+      projects = Array.wrap(projects)
+      folders = Array.wrap(folders)
+
+      [].tap do |roles|
+        roles << { 'type' => 'admin' } if admin
+        projects.each { |project| roles << { 'type' => 'project_moderator', 'project_id' => project.id } }
+        folders.each { |folder| roles << { 'type' => 'project_folder_moderator', 'project_folder_id' => folder.id } }
+      end
+    end
+
+    context 'when the roles are not redundant' do
+      where(:admin?, :projects, :folders) do
+        [
+          [true, nil, nil],
+          [true, ref(:another_project), nil],
+          [true, nil, ref(:folder)],
+          [false, ref(:another_project), ref(:folder)],
+          [false, ref(:projects_in_folder), nil],
+          [false, nil, [ref(:folder), ref(:another_folder)]],
+          [false, ref(:project_in_another_folder), ref(:folder)]
+        ]
+      end
+
+      with_them do
+        it 'does not modify the roles' do
+          user.roles = create_roles(projects, folders, admin: admin?)
+          expect(user.compacted_roles).to match(user.roles)
+        end
+      end
+    end
+
+    context 'when the roles are redundant' do
+      it 'removes redundant roles (1)' do
```

Comment:
I wonder if we can make these it tests more descriptive. It takes some 🧠 power to grasp the tests.

E.g. "it removes project roles when the user has moderator roles for a folder and projects inside the folder" instead of "removes redundant roles (1)".

Reply:
Probably, but personally parsing the sentence takes me as much brain power as parsing the code. I thought I managed to make the tests quite concise and pretty self-explanatory 😅